aten/src/ATen/native/nested/NestedTensorMath.cpp

```diff
@@ -905,6 +1010,22 @@ Tensor matmul_nested(const Tensor& self, const Tensor& mat2) {
     self_dim_size,
     "second last dimension of mat2 has sizes",
     mat2_dim_size);
+
+  // use bmm inference-only fast path for [N, n_heads, *, head_dim] [N, n_heads, head_dim, *]
+  if (self_dim == 4 && mat2_dim == 4) {
```

Comment:
You can check here whether the inputs require gradients and then branch on that

Reply:
In particular you can check

```
torch.is_grad_enabled() and any(x.requires_grad for x in tensor_args):
```

From C++ that's [`GradMode::is_enabled()`](https://github.com/pytorch/pytorch/blob/a67621a6ca19b0c1423a7b136bfd90d8f04182fb/torch/csrc/autograd/init.cpp#L516) for torch.is_grad_enabled and `x.requires_grad()`